### PR TITLE
[opinfo] use dtypes instead of dtypesIfCPU

### PR DESCRIPTION
**Pull request by kshitij12345**
Replace usage of `dtypesIfCPU` with `dtypes` in OpInfo class and also make it a mandatory argument.

Also added DeprecationWarning on using `dtypesIfCPU`

This raises a question :
For an OpInfo entry, currently `dtypes` works for any external backend, `dtypesIfCPU` for CPU and `dtypesIfCUDA` and `dtypesIfROCM` for CUDA and ROCm respectively.

If we merge `dtypes` and `dtypesIfCPU`, then for cases where external backend `dtypes` don't match cpu `dtypes` then it will lead to failures.

Currently there are few issues (5 failures) due to this on XLA (we may add relevant skips for the same). If we agree that skip should be added, then should it be added via OpInfo using decorators mechanism or at the XLA end? I think XLA end makes more sense to me to have one source of skips.

<details>

<summary>XLA Fail Log</summary>

```
Nov 01 11:48:26 ======================================================================
Nov 01 11:48:26 ERROR [0.016s]: test_reference_eager_histogram_xla_float32 (__main__.TestOpInfoXLA)
Nov 01 11:48:26 ----------------------------------------------------------------------
Nov 01 11:48:26 Traceback (most recent call last):
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
Nov 01 11:48:26     result = test(self, **param_kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
Nov 01 11:48:26     return test(*args, **kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 411, in test_reference_eager
Nov 01 11:48:26     self.compare_with_eager_reference(op, sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 397, in compare_with_eager_reference
Nov 01 11:48:26     cpu_inp, cpu_args, cpu_kwargs = cpu(sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 393, in cpu
Nov 01 11:48:26     sample.args), to_cpu(sample.kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 386, in to_cpu
Nov 01 11:48:26     return {k: to_cpu(v) for k, v in x.items()}
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 386, in <dictcomp>
Nov 01 11:48:26     return {k: to_cpu(v) for k, v in x.items()}
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 390, in to_cpu
Nov 01 11:48:26     raise ValueError("Unknown type {0}!".format(type(x)))
Nov 01 11:48:26 ValueError: Unknown type <class 'NoneType'>!
Nov 01 11:48:26 
Nov 01 11:48:26 ======================================================================
Nov 01 11:48:26 FAIL [0.575s]: test_reference_eager___rmatmul___xla_int64 (__main__.TestOpInfoXLA)
Nov 01 11:48:26 ----------------------------------------------------------------------
Nov 01 11:48:26 Traceback (most recent call last):
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
Nov 01 11:48:26     result = test(self, **param_kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
Nov 01 11:48:26     return test(*args, **kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 411, in test_reference_eager
Nov 01 11:48:26     self.compare_with_eager_reference(op, sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 402, in compare_with_eager_reference
Nov 01 11:48:26     self.assertEqual(actual, expected, exact_dtype=True, exact_device=False)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 607, in assertEqual
Nov 01 11:48:26     return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1903, in assertEqual
Nov 01 11:48:26     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Nov 01 11:48:26 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0.001 and atol=0.001, found 44 element(s) (out of 50) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 9.187201950435738e+18 (-9.187201950435738e+18 vs. 34.0), which occurred at index (0, 4).
Nov 01 11:48:26 
Nov 01 11:48:26 ======================================================================
Nov 01 11:48:26 FAIL [0.137s]: test_reference_eager_linalg_multi_dot_xla_int64 (__main__.TestOpInfoXLA)
Nov 01 11:48:26 ----------------------------------------------------------------------
Nov 01 11:48:26 Traceback (most recent call last):
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
Nov 01 11:48:26     result = test(self, **param_kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
Nov 01 11:48:26     return test(*args, **kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 411, in test_reference_eager
Nov 01 11:48:26     self.compare_with_eager_reference(op, sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 402, in compare_with_eager_reference
Nov 01 11:48:26     self.assertEqual(actual, expected, exact_dtype=True, exact_device=False)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 607, in assertEqual
Nov 01 11:48:26     return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1903, in assertEqual
Nov 01 11:48:26     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Nov 01 11:48:26 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0.001 and atol=0.001, found 4 element(s) (out of 4) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 140230883884432.0 (0.0 vs. 140230883884432.0), which occurred at index (0, 0).
Nov 01 11:48:26 
Nov 01 11:48:26 ======================================================================
Nov 01 11:48:26 FAIL [0.461s]: test_reference_eager_matmul_xla_int64 (__main__.TestOpInfoXLA)
Nov 01 11:48:26 ----------------------------------------------------------------------
Nov 01 11:48:26 Traceback (most recent call last):
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
Nov 01 11:48:26     result = test(self, **param_kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
Nov 01 11:48:26     return test(*args, **kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 411, in test_reference_eager
Nov 01 11:48:26     self.compare_with_eager_reference(op, sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 402, in compare_with_eager_reference
Nov 01 11:48:26     self.assertEqual(actual, expected, exact_dtype=True, exact_device=False)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 607, in assertEqual
Nov 01 11:48:26     return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1903, in assertEqual
Nov 01 11:48:26     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Nov 01 11:48:26 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=0.001 and atol=0.001, found 37 element(s) (out of 50) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 7.661375630332297e+18 (-7.66128151259864e+18 vs. 94117733658072.0), which occurred at index (4, 5).
Nov 01 11:48:26 
Nov 01 11:48:26 ======================================================================
Nov 01 11:48:26 FAIL [0.050s]: test_reference_eager_remainder_autodiffed_xla_int64 (__main__.TestOpInfoXLA)
Nov 01 11:48:26 ----------------------------------------------------------------------
Nov 01 11:48:26 Traceback (most recent call last):
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
Nov 01 11:48:26     result = test(self, **param_kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
Nov 01 11:48:26     return test(*args, **kwargs)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 411, in test_reference_eager
Nov 01 11:48:26     self.compare_with_eager_reference(op, sample_input)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/test_ops.py", line 402, in compare_with_eager_reference
Nov 01 11:48:26     self.assertEqual(actual, expected, exact_dtype=True, exact_device=False)
Nov 01 11:48:26   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 607, in assertEqual
Nov 01 11:48:26     return DeviceTypeTestBase.assertEqual(self, x, y, *args, **kwargs)
Nov 01 11:48:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1903, in assertEqual
Nov 01 11:48:26     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Nov 01 11:48:26 AssertionError: False is not true : Tensors failed to compare as equal!Attempted to compare equality of tensors with different dtypes. Got dtypes torch.int64 and torch.float32.
Nov 01 11:48:26 
Nov 01 11:48:26 ----------------------------------------------------------------------
```

</details>

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -485,10 +485,7 @@ def __init__(self,
                  sample_inputs_func=None,  # function to generate sample inputs
 
                  # the following metadata relates to dtype support and is tested for correctness in test_ops.py
-                 dtypes=floating_types(),  # dtypes this function is expected to work with
-                 # the following dtypesIf... options override the dtypes value
-                 # on their respective device types
-                 dtypesIfCPU=None,  # dtypes this function is expected to work with on CPU
+                 dtypes=None,  # dtypes this function is expected to work with on CPU
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:488]: Now defaults to `None`

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -548,7 +545,8 @@ def __init__(self,
                  assert_jit_shape_analysis=False,  # assert that jit shape analysis fully propagates shape
                  ):
 
-        dtypes_args = (dtypes, dtypesIfCPU, dtypesIfCUDA, dtypesIfROCM)
+        assert dtypes is not None
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:548]: Make passing `dtype` mandatory.

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

**Comment by mruberry** [comment]
This is really cool, and thanks for the ping, @kshitij12345!

So here's what I'm thinking:

- let's leave `dtypesIfCPU` on OpInfos and as a decorator, however...
- let's update `dtypes` to be `dtypesIfCPU` wherever possible to get rid of `dtypesIfCPU` uses
- let's require `@dtypes` be specified if `dtypesIf...` is specified for any device type (this is not part of the current PR)
- for XLA let's use `@skipXLA` and add (and use) `dtypesIfXLA` to handle its issues
- we can probably replace some uses of `@onlyNativeDeviceTypes` with `@skipXLA`

This does a few things:

- engineers are confused today about `dtypes` because it's so abstract -- what is the perfect set of abstract dtypes? This makes the guidance "whatever the CPU uses", which should be straightforward
- It keeps `dtypesIfCPU` just in case and for symmetry (originally I know we wanted to eliminate it, but some engineers expressed concern about just killing it so let's appease them by leaving it but not using it)
- It means this PR just needs a tweak to preserve `dtypesIfCPU` but update the direction for how to use `dtypes`
- This PR can also handle the XLA failures by adding `dtypesIfXLA`, I think? We should NOT verify those dtypes are correct or extend XLA testing when we do this

So basically this PR is 95% perfect, just a few tweaks to keep `dtypesIfCPU` and add `dtypesIfXLA`, and in the future it'd be nice to require the `@dtypes` decorator be used if a device-specific decorator is used with a test.

Does that make sense? What are your thoughts?

**Comment by kshitij12345** [comment]
This in general makes sense. Though I am a bit skeptical about `dtypesIfXLA`. How about `dtypesIfNonNative` (anything except our current native devices). And external backend libraries like XLA can update that field before instantiating the test if required?

**Comment by mruberry** [comment]
> This in general makes sense. Though I am a bit skeptical about `dtypesIfXLA`. How about `dtypesIfNonNative` (anything except our current native devices). And external backend libraries like XLA can update that field before instantiating the test if required?
> 
> Does it make sense? What do you think about that? Thanks!

`dtypesIfNonNative` sounds good to me

**Comment by kshitij12345** [comment]
@mruberry 
Have added `dtypesIfCPU` back (though it is not used by any OpInfo now). Also have added a DeprecationWarning when `dtypesIfCPU` is passed.

For now have added `expectedFailure` on XLA. (as the failures are due to the new dtypes being added for XLA). XLA CI passes now.

I am planning to do the following in seperate follow-up PRs
* Add `dtypesIfNonNative`
* Updating the code so that @dtypes is mandatory where @dtypesIf is used (Have a local patch, would like to discuss about testing and error message offline)

With this the PR should be ready. Let me know if I missed anything.

Thanks!

**Comment by kshitij12345** [comment]
@mruberry I think this should be (almost) ready. Thanks

File: test/test_testing.py [hunk 3]
```diff
@@ -598,7 +598,7 @@ def test_get_supported_dtypes(self, device):
             dynamic_dtypes = opinfo_helper.get_supported_dtypes(op.op, op.sample_inputs_func, self.device_type)
             dynamic_dispatch = opinfo_helper.dtypes_dispatch_hint(dynamic_dtypes)
             if self.device_type == 'cpu':
-                dtypes = op.dtypesIfCPU
+                dtypes = op.dtypes
```
Comment (mruberry) [inline test/test_testing.py:601]: Shouldn't this go back to being dtypesIfCPU in case someone does set it?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -548,7 +549,9 @@ def __init__(self,
                  assert_jit_shape_analysis=False,  # assert that jit shape analysis fully propagates shape
                  ):
 
-        dtypes_args = (dtypes, dtypesIfCPU, dtypesIfCUDA, dtypesIfROCM)
+        assert dtypes is not None, ("`dtypes` argument for OpInfo should not be None,",
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:552]: Nice assertion

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -548,7 +549,9 @@ def __init__(self,
...
+                                    f"but found None for {name}")
+        dtypes_args = (dtypes, dtypesIfCUDA, dtypesIfROCM)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:554]: dtypesIfCPU should still be here, right?

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -592,6 +595,8 @@ def __init__(self,
             else dtypesIfCUDA if dtypesIfCUDA is not None
             else dtypes)
 
+        if dtypesIfCPU is not None:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:598]: Let's just kill this warning for now; I don't think people will use it but if they do that's fine

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -485,7 +486,7 @@ def __init__(self,
                  sample_inputs_func=None,  # function to generate sample inputs
 
                  # the following metadata relates to dtype support and is tested for correctness in test_ops.py
-                 dtypes=floating_types(),  # dtypes this function is expected to work with
+                 dtypes=None,  # dtypes this function is expected to work with
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:489]: Let's update this comment to be "dtypes this function is expected to work with on the CPU"

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -485,7 +486,7 @@ def __init__(self,
...
                  # the following dtypesIf... options override the dtypes value
                  # on their respective device types
                  dtypesIfCPU=None,  # dtypes this function is expected to work with on CPU
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:492]: Let's update this comment to be "dtypes this function is expected to work with on the CPU, typically unnecessary since it's (now) redundant with the dtypes kwarg above"

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -8184,18 +8153,24 @@ def ref_pairwise_distance(input1, input2):
...
            ref=np.remainder,
            variant_test_name='autodiffed',
-           dtypesIfCPU=all_types_and(torch.float16, torch.bool, torch.bfloat16),
+           dtypes=all_types_and(torch.float16, torch.bool, torch.bfloat16),
            dtypesIfCUDA=all_types_and(torch.float16, torch.bool, torch.bfloat16),
            supports_forward_ad=True,
            assert_autodiffed=True,
-           sample_inputs_func=partial(sample_inputs_fmod_remainder, autodiffed=True)),
+           sample_inputs_func=partial(sample_inputs_fmod_remainder, autodiffed=True),
+           decorators=(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:8168]: Nice update to xfails

File: torch/testing/_internal/opinfo_helper.py [hunk 10]
```diff
@@ -129,11 +129,11 @@ def is_dynamic_dtype_set(op):
 def str_format_dynamic_dtype(op):
     fmt_str = """
         OpInfo({name},
-               dtypes={dtypesIfCPU},
+               dtypes={dtypes},
```
Comment (mruberry) [inline torch/testing/_internal/opinfo_helper.py:132]: Cool update here

**Review by mruberry — APPROVED** [review]
Just a few small suggestions inline and then this should be good to go!

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:489]: Will update it but a comment : This is still generic for all devices (if their respective fields are not set right), so I think the existing `dtypes this function is expected to work makes ` sense. 🤔

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py]: That's a good point; what about merging our ideas with something like "dtypes this function works with on the CPU, inherited by other device types that don't specify their own dtypes"?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py]: That sounds good. Thanks!

**Comment by kshitij12345** [comment]
@mruberry CI failures look irrelevant.

**Comment by kshitij12345** [comment]
@mruberry 

Functorch tests would fail post this PR as it specifies few `OpInfo` without passing `dtypes`

https://github.com/pytorch/functorch/blob/f16519db99cdc945a9451f8073f708ddcf2c1a09/test/functorch_additional_op_db.py#L79-L94

**Comment by mruberry** [comment]
> @mruberry `functorch` tests would fail at collection post this PR as it specifies few `OpInfo` without passing `dtypes`
> 
> https://github.com/pytorch/functorch/blob/f16519db99cdc945a9451f8073f708ddcf2c1a09/test/functorch_additional_op_db.py#L79-L94
> 
> cc: @zou3519

Good spot, @kshitij12345. @zou3519 @Chillee maybe we can merge those extra OpInfos now? The workaround for this PR should take all of 10 seconds to add (just add a generic `dtypes`).

**Comment by zou3519** [comment]
>  @zou3519 @Chillee maybe we can merge those extra OpInfos now? The workaround for this PR should take all of 10 seconds to add (just add a generic dtypes).

Are you asking if we can merge this PR (#67619) and fix the problems later, or if we can upstream https://github.com/pytorch/functorch/blob/f16519db99cdc945a9451f8073f708ddcf2c1a09/test/functorch_additional_op_db.py#L79-L94 into PyTorch core?

For the former, yes, feel free to merge this PR first and we will the problems on our end later.

For the latter, a lot of those have actually been merged into PyTorch core and we haven't gotten around to deleting them from functorch

**Comment by mruberry** [comment]
> > @zou3519 @Chillee maybe we can merge those extra OpInfos now? The workaround for this PR should take all of 10 seconds to add (just add a generic dtypes).
> 
> Are you asking if we can merge this PR (#67619) and fix the problems later, or if we can upstream https://github.com/pytorch/functorch/blob/f16519db99cdc945a9451f8073f708ddcf2c1a09/test/functorch_additional_op_db.py#L79-L94 into PyTorch core?
> 
> For the former, yes, feel free to merge this PR first and we will the problems on our end later.
> 
> For the latter, a lot of those have actually been merged into PyTorch core and we haven't gotten around to deleting them from functorch

I thought you might want to merge a PR to functorch before this went in, actually; it'd be very simple to do. But sure, we won't block this PR.
